deps/rabbit/src/rabbit_msg_store.erl

```diff
@@ -518,14 +523,264 @@ read(MsgId,
+            %% @todo It's probably a bug if we don't get a positive ref count.
             case index_lookup_positive_ref_count(MsgId, CState) of
                 not_found   -> Defer();
-                MsgLocation -> client_read1(MsgLocation, Defer, CState)
+                MsgLocation -> client_read2(MsgLocation, Defer, CState)
             end;
         [{MsgId, Msg, _CacheRefCount}] ->
             {{ok, Msg}, CState}
     end.
 
+%% @todo read_many
+%% @todo We can only read_many when the backend index is rabbit_msg_store_ets_index
+%%       since we rely on ets:select to get the index entries.
+
+%% @todo OK first we need to do the ets:lookups to get the cached values.
+%%       There's probably a way to optimize this to avoid lookups for
+%%       messages that are older: since the cache will be generational
+%%       we can simply stop looking in the cache at the first MsgId
+%%       that doesn't find anything, AS LONG AS the MsgIds are ordered
+%%       from younger to older.
+%% @todo OK but for fan-out it's theoretically possible that a message
+%%       gets flushed from the cache but sits in a different position
+%%       in the queue. That means we may end up with some extra cache misses
+%%       but nothing that should be too noticeable.
+%%
+%%       Then with what remains, we do an index_lookup_positive_ref_count/2
+%%       of the first MsgId to read. If it is not found (bug?), ignore and
+%%       try the next.
+%% @todo This means we may not return as many messages as asked and
+%%       therefore the rabbit_variable_queue implementation needs
+%%       to be updated.
+%%
+%%       Once we have a location, we first mark ourselves as a reader.
+%%       We then mark the file handle open and get the index again,
+%%       only this time we do an ets:select to get all relevant
+%%       messages for this file. Since we get the messages for the
+%%       file we either find locations for the file or nothing
+%%       (when compacting we first update the index then we do
+%%       the file changes if possible). If we find nothing, we go
+%%       back at the start of the loop and retry for the same MsgId
+%%       as before. Otherwise we can continue.
+%%
+%%       We then do the pread and continue with the remaining MsgIds
+%%       that were not found in the MsgLocations we got. We eventually
+%%       return whatever we managed to read.
+
+%% @todo First I will get an implementation that simply calls
+%%       read many times in order to ensure the rest of the changes
+%%       don't break anything.
+
+-spec read_many([rabbit_types:msg_id()], client_msstate()) -> [msg()]. %% @todo Spec of return
+
+%% We disable read_many when the index module is not ETS.
+read_many(_, #client_msstate{ index_module = IndexMod })
+        when IndexMod =/= rabbit_msg_store_ets_index ->
+    [];
+read_many(MsgIds, CState) ->
+% @todo    file_handle_cache_stats:update(msg_store_read),
```

Comment:
Need to figure out what to do about this one. For my purposes a different stat would help of course, but I think we need to add the N reads we do here to this stat.